File: resources/engine/cfg.py
# -*- coding: utf-8 -*-
"""
Построение графа потока управления (CFG) из линейного списка инструкций:
базовые блоки, рёбра, дерево доминаторов, естественные циклы (back edges).
"""


class Block:
    __slots__ = ("start", "end", "instrs", "succs", "preds", "idom",
                 "domfrontier", "kind_hint", "handler_types")

    def __init__(self, start):
        self.start = start
        self.end = None            # pc после последней инструкции блока (exclusive)
        self.instrs = []           # list[Instruction]
        self.succs = []            # list[int] (start pc of successor blocks)
        self.preds = []            # list[int]
        self.idom = None
        self.domfrontier = set()
        self.kind_hint = None      # для аннотаций (напр. 'handler')
        self.handler_types = []    # если это начало exception-handler'а: [(catch_type, ExceptionEntry), ...]

    def __repr__(self):
        return f"<Block {self.start}..{self.end}>"


class CFG:
    def __init__(self, instrs, order, exceptions):
        self.instrs = instrs
        self.order = order
        self.exceptions = exceptions
        self.blocks = {}       # start_pc -> Block
        self.entry = order[0] if order else None
        self._build()
        if self.entry is not None:
            self._compute_dominators()

    def _leaders(self):
        leaders = set()
        if self.order:
            leaders.add(self.order[0])
        for pc in self.order:
            ins = self.instrs[pc]
            if ins.is_branch:
                leaders.add(ins.target)
                leaders.add(ins.next_pc)
            elif ins.is_switch:
                for v, t in ins.targets.items():
                    leaders.add(t)
                leaders.add(ins.next_pc)
            elif ins.is_unconditional:
                leaders.add(ins.next_pc)
        # exception handlers and try-range starts are also leaders
        for e in self.exceptions:
            leaders.add(e.start_pc)
            leaders.add(e.end_pc)
            leaders.add(e.handler_pc)
        leaders = {pc for pc in leaders if pc in self.instrs}
        return leaders

    def _build(self):
        leaders = sorted(self._leaders())
        if not leaders:
            return
        for i, start in enumerate(leaders):
            b = Block(start)
            end = leaders[i + 1] if i + 1 < len(leaders) else (self.order[-1] and self.instrs[self.order[-1]].next_pc)
            pc = start
            while pc is not None and pc < end and pc in self.instrs:
                ins = self.instrs[pc]
                b.instrs.append(ins)
                pc = ins.next_pc
            b.end = pc
            self.blocks[start] = b

        starts = sorted(self.blocks)
        for i, start in enumerate(starts):
            b = self.blocks[start]
            if not b.instrs:
                continue
            last = b.instrs[-1]
            if last.is_switch:
                for v, t in last.targets.items():
                    if t in self.blocks:
                        b.succs.append(t)
            elif last.mnemonic in ("goto", "goto_w"):
                if last.target in self.blocks:
                    b.succs.append(last.target)
            elif last.is_conditional:
                if last.target in self.blocks:
                    b.succs.append(last.target)
                if last.next_pc in self.blocks:
                    b.succs.append(last.next_pc)
            elif last.is_return or last.mnemonic == "athrow":
                pass  # no fallthrough successor
            else:
                if i + 1 < len(starts):
                    b.succs.append(starts[i + 1])
        # exception edges: block covering try range -> handler (approx: any block
        # inside [start_pc,end_pc) gets an edge to handler_pc)
        for e in self.exceptions:
            if e.handler_pc not in self.blocks:
                continue
            for start in starts:
                if e.start_pc <= start < e.end_pc:
                    self.blocks[start].succs.append(e.handler_pc)
                    hb = self.blocks[e.handler_pc]
                    hb.handler_types.append((e.catch_type, e))

        for start, b in self.blocks.items():
            for s in b.succs:
                if start not in self.blocks[s].preds:
                    self.blocks[s].preds.append(start)

    # ---------------- dominators (iterative, simple) ----------------

    def _compute_dominators(self):
        starts = sorted(self.blocks)
        idx = {s: i for i, s in enumerate(starts)}
        # reverse postorder
        rpo = self._reverse_postorder()
        rpo_index = {s: i for i, s in enumerate(rpo)}
        idom = {self.entry: self.entry}
        changed = True
        while changed:
            changed = False
            for b in rpo:
                if b == self.entry:
                    continue
                preds = [p for p in self.blocks[b].preds if p in idom]
                if not preds:
                    continue
                new_idom = preds[0]
                for p in preds[1:]:
                    new_idom = self._intersect(new_idom, p, idom, rpo_index)
                if idom.get(b) != new_idom:
                    idom[b] = new_idom
                    changed = True
        for b in self.blocks:
            self.blocks[b].idom = idom.get(b)
        self.idom = idom

    def _intersect(self, a, b, idom, rpo_index):
        while a != b:
            while rpo_index[a] > rpo_index[b]:
                a = idom[a]
            while rpo_index[b] > rpo_index[a]:
                b = idom[b]
        return a

    def _reverse_postorder(self):
        visited = set()
        order = []
        def dfs(n):
            visited.add(n)
            for s in self.blocks[n].succs:
                if s not in visited and s in self.blocks:
                    dfs(s)
            order.append(n)
        if self.entry in self.blocks:
            dfs(self.entry)
        order.reverse()
        # append unreachable blocks (dead code) at the end, in address order,
        # so nothing is silently dropped
        for s in sorted(self.blocks):
            if s not in visited:
                order.append(s)
        return order

    def dominates(self, a, b):
        """a доминирует над b?"""
        n = b
        seen = 0
        while True:
            if n == a:
                return True
            if n == self.idom.get(n) or n not in self.idom:
                return n == a
            if seen > len(self.blocks) + 2:
                return False
            n = self.idom[n]
            seen += 1

    def natural_loops(self):
        """Возвращает список (header, set(body_blocks), back_edge_sources)."""
        loops = {}
        for start, b in self.blocks.items():
            for s in b.succs:
                if s in self.blocks and self.dominates(s, start):
                    # back edge start -> s ; s is loop header
                    loops.setdefault(s, set()).add(start)
        result = []
        for header, tails in loops.items():
            body = {header}
            stack = list(tails)
            while stack:
                n = stack.pop()
                if n not in body:
                    body.add(n)
                    for p in self.blocks[n].preds:
                        if p not in body:
                            stack.append(p)
            result.append((header, body, tails))
        return result

    def reverse_postorder_list(self):
        return self._reverse_postorder()

    # ---------------- post-dominators (for if/else merge-point detection) ----------------

    def compute_postdominators(self):
        """ipdom[b] = ближайший постдоминатор b (адрес блока, либо EXIT-маркер None
        если ни один реальный блок не является общим постдоминатором - т.е. все пути
        из b завершаются return/throw без общей точки схождения)."""
        EXIT = "__EXIT__"
        preds_rev = {b: list(self.blocks[b].succs) for b in self.blocks}   # reversed succ = pred
        succs_rev = {b: list(self.blocks[b].preds) for b in self.blocks}   # reversed pred = succ
        exit_preds = []
        for b, blk in self.blocks.items():
            if not blk.succs:
                exit_preds.append(b)
        # rpo of reversed graph starting from EXIT
        visited = set()
        order = []
        def dfs(n):
            visited.add(n)
            nxt = exit_preds if n == EXIT else succs_rev.get(n, [])
            for s in nxt:
                if s not in visited:
                    dfs(s)
            order.append(n)
        dfs(EXIT)
        for b in sorted(self.blocks):
            if b not in visited:
                order.append(b)
                visited.add(b)
        order.reverse()
        rpo_index = {n: i for i, n in enumerate(order)}

        idom = {EXIT: EXIT}
        changed = True
        while changed:
            changed = False
            for b in order:
                if b == EXIT:
                    continue
                ps = preds_rev.get(b, [])
                if b in exit_preds:
                    ps = ps + [EXIT]
                ps = [p for p in ps if p in idom]
                if not ps:
                    continue
                new_idom = ps[0]
                for p in ps[1:]:
                    new_idom = self._intersect_generic(new_idom, p, idom, rpo_index)
                if idom.get(b) != new_idom:
                    idom[b] = new_idom
                    changed = True
        result = {}
        for b in self.blocks:
            ip = idom.get(b)
            result[b] = None if ip == EXIT or ip is None else ip
        self.ipdom = result
        return result

    def _intersect_generic(self, a, b, idom, rpo_index):
        while a != b:
            while rpo_index[a] > rpo_index[b]:
                a = idom[a]
            while rpo_index[b] > rpo_index[a]:
                b = idom[b]
        return a

File: resources/engine_cpp/src/cfg.cpp
// cfg.cpp - см. cfg.hpp. 1:1 порт cfg.py (включая воспроизведённые
// особенности оригинала - см. HANDOFF_28).
#include "cfg.hpp"

#include <algorithm>
#include <functional>

namespace nd {

namespace {
constexpr int64_t EXIT_MARKER = -1;  // аналог строкового сентинела "__EXIT__" в Python
}

CFG::CFG(const std::map<size_t, Instruction>& instrs, const std::vector<size_t>& order,
         const std::vector<ExceptionEntry>& exceptions) {
    for (auto& [pc, ins] : instrs) instrs_[static_cast<int64_t>(pc)] = ins;
    for (size_t pc : order) order_.push_back(static_cast<int64_t>(pc));
    exceptions_ = exceptions;
    entry = order_.empty() ? std::nullopt : std::optional<int64_t>(order_.front());
    build();
    if (entry.has_value()) compute_dominators();
}

std::set<int64_t> CFG::leaders() const {
    std::set<int64_t> ldrs;
    if (!order_.empty()) ldrs.insert(order_.front());
    for (int64_t pc : order_) {
        const Instruction& ins = instrs_.at(pc);
        if (ins.is_branch()) {
            if (ins.target.has_value()) ldrs.insert(*ins.target);
            ldrs.insert(static_cast<int64_t>(ins.next_pc));
        } else if (ins.is_switch()) {
            if (ins.targets.has_value()) {
                for (auto& [v, t] : *ins.targets) ldrs.insert(t);
            }
            ldrs.insert(static_cast<int64_t>(ins.next_pc));
        } else if (ins.is_unconditional()) {
            ldrs.insert(static_cast<int64_t>(ins.next_pc));
        }
    }
    for (auto& e : exceptions_) {
        ldrs.insert(e.start_pc);
        ldrs.insert(e.end_pc);
        ldrs.insert(e.handler_pc);
    }
    std::set<int64_t> out;
    for (int64_t pc : ldrs) {
        if (instrs_.count(pc)) out.insert(pc);
    }
    return out;
}

void CFG::build() {
    std::set<int64_t> leaders_set = leaders();
    if (leaders_set.empty()) return;
    std::vector<int64_t> leaders_vec(leaders_set.begin(), leaders_set.end());  // std::set уже отсортирован

    for (size_t i = 0; i < leaders_vec.size(); ++i) {
        int64_t start = leaders_vec[i];
        Block b;
        b.start = start;
        int64_t end;
        if (i + 1 < leaders_vec.size()) {
            end = leaders_vec[i + 1];
        } else {
            // Зеркалит Python: `self.order[-1] and self.instrs[self.order[-1]].next_pc`
            // - "and"-трюк даёт 0 (а НЕ next_pc), если order[-1] == 0 (метод из
            // ОДНОЙ инструкции на pc=0). Это особенность/баг оригинала -
            // воспроизведена намеренно, не исправлена. См. HANDOFF_28.
            int64_t last = order_.back();
            end = (last == 0) ? 0 : static_cast<int64_t>(instrs_.at(last).next_pc);
        }
        int64_t pc = start;
        while (pc < end && instrs_.count(pc)) {
            b.instrs.push_back(instrs_.at(pc));
            pc = static_cast<int64_t>(instrs_.at(pc).next_pc);
        }
        b.end = pc;
        blocks[start] = std::move(b);
    }

    std::vector<int64_t> starts;
    for (auto& [s, _] : blocks) starts.push_back(s);  // std::map -> уже отсортировано

    for (size_t i = 0; i < starts.size(); ++i) {
        int64_t start = starts[i];
        Block& b = blocks[start];
        if (b.instrs.empty()) continue;
        const Instruction& last = b.instrs.back();
        if (last.is_switch()) {
            if (last.targets.has_value()) {
                for (auto& [v, t] : *last.targets) {
                    if (blocks.count(t)) b.succs.push_back(t);
                }
            }
        } else if (last.mnemonic == "goto" || last.mnemonic == "goto_w") {
            if (last.target.has_value() && blocks.count(*last.target)) b.succs.push_back(*last.target);
        } else if (last.is_conditional()) {
            if (last.target.has_value() && blocks.count(*last.target)) b.succs.push_back(*last.target);
            if (blocks.count(static_cast<int64_t>(last.next_pc))) b.succs.push_back(static_cast<int64_t>(last.next_pc));
        } else if (last.is_return() || last.mnemonic == "athrow") {
            // нет fallthrough-преемника
        } else {
            if (i + 1 < starts.size()) b.succs.push_back(starts[i + 1]);
        }
    }

    // Рёбра исключений: см. предупреждение в cfg.hpp - дублирующиеся записи
    // handler_types воспроизведены намеренно точно как в оригинале.
    for (auto& e : exceptions_) {
        if (!blocks.count(e.handler_pc)) continue;
        for (int64_t start : starts) {
            if (e.start_pc <= start && start < e.end_pc) {
                blocks[start].succs.push_back(e.handler_pc);
                Block& hb = blocks[e.handler_pc];
                hb.handler_types.emplace_back(e.catch_type, e);
            }
        }
    }

    for (auto& [start, b] : blocks) {
        for (int64_t s : b.succs) {
            auto& target_preds = blocks[s].preds;
            if (std::find(target_preds.begin(), target_preds.end(), start) == target_preds.end()) {
                target_preds.push_back(start);
            }
        }
    }
}

// ---------------- dominators ----------------

void CFG::compute_dominators() {
    std::vector<int64_t> rpo = reverse_postorder();
    std::map<int64_t, size_t> rpo_index;
    for (size_t i = 0; i < rpo.size(); ++i) rpo_index[rpo[i]] = i;

    std::map<int64_t, int64_t> idom;
    idom[*entry] = *entry;
    bool changed = true;
    while (changed) {
        changed = false;
        for (int64_t b : rpo) {
            if (b == *entry) continue;
            std::vector<int64_t> preds;
            for (int64_t p : blocks[b].preds) {
                if (idom.count(p)) preds.push_back(p);
            }
            if (preds.empty()) continue;
            int64_t new_idom = preds[0];
            for (size_t k = 1; k < preds.size(); ++k) {
                int64_t a = new_idom, c = preds[k];
                while (a != c) {
                    while (rpo_index[a] > rpo_index[c]) a = idom[a];
                    while (rpo_index[c] > rpo_index[a]) c = idom[c];
                }
                new_idom = a;
            }
            auto it = idom.find(b);
            if (it == idom.end() || it->second != new_idom) {
                idom[b] = new_idom;
                changed = true;
            }
        }
    }
    for (auto& [b, blk] : blocks) {
        auto it = idom.find(b);
        blk.idom = (it != idom.end()) ? std::optional<int64_t>(it->second) : std::nullopt;
    }
    idom_ = std::move(idom);
}

std::vector<int64_t> CFG::reverse_postorder() const {
    std::set<int64_t> visited;
    std::vector<int64_t> order;
    std::function<void(int64_t)> dfs = [&](int64_t n) {
        visited.insert(n);
        auto it = blocks.find(n);
        if (it != blocks.end()) {
            for (int64_t s : it->second.succs) {
                if (!visited.count(s) && blocks.count(s)) dfs(s);
            }
        }
        order.push_back(n);
    };
    if (entry.has_value() && blocks.count(*entry)) dfs(*entry);
    std::reverse(order.begin(), order.end());
    for (auto& [s, _] : blocks) {  // std::map -> сортированный порядок, как sorted(self.blocks)
        if (!visited.count(s)) {
            order.push_back(s);
            visited.insert(s);
        }
    }
    return order;
}

std::vector<int64_t> CFG::reverse_postorder_list() const { return reverse_postorder(); }

bool CFG::dominates(int64_t a, int64_t b) const {
    int64_t n = b;
    size_t seen = 0;
    while (true) {
        if (n == a) return true;
        auto it = idom_.find(n);
        if (it == idom_.end()) return n == a;
        if (it->second == n) return n == a;
        if (seen > blocks.size() + 2) return false;
        n = it->second;
        seen += 1;
    }
}

std::vector<std::tuple<int64_t, std::set<int64_t>, std::set<int64_t>>> CFG::natural_loops() const {
    // Порядок заголовков (header) в результирующем списке ДОЛЖЕН совпадать
    // с порядком их ПЕРВОГО обнаружения (как insertion-order Python dict) -
    // воспроизведено явным вектором first-seen порядка.
    std::vector<int64_t> header_order;
    std::map<int64_t, std::set<int64_t>> loops;
    for (auto& [start, b] : blocks) {  // sorted(self.blocks.items()) по построению std::map
        for (int64_t s : b.succs) {
            if (blocks.count(s) && dominates(s, start)) {
                auto [it, inserted] = loops.try_emplace(s);
                if (inserted) header_order.push_back(s);
                it->second.insert(start);
            }
        }
    }
    std::vector<std::tuple<int64_t, std::set<int64_t>, std::set<int64_t>>> result;
    for (int64_t header : header_order) {
        const std::set<int64_t>& tails = loops[header];
        std::set<int64_t> body;
        body.insert(header);
        std::vector<int64_t> stack(tails.begin(), tails.end());  // порядок стека не влияет на итоговое множество body
        while (!stack.empty()) {
            int64_t n = stack.back();
            stack.pop_back();
            if (!body.count(n)) {
                body.insert(n);
                auto it = blocks.find(n);
                if (it != blocks.end()) {
                    for (int64_t p : it->second.preds) {
                        if (!body.count(p)) stack.push_back(p);
                    }
                }
            }
        }
        result.emplace_back(header, std::move(body), tails);
    }
    return result;
}

// ---------------- post-dominators ----------------

std::map<int64_t, std::optional<int64_t>> CFG::compute_postdominators() {
    std::map<int64_t, std::vector<int64_t>> preds_rev;  // succs of b (используется как "предки" в развёрнутом графе)
    std::map<int64_t, std::vector<int64_t>> succs_rev;  // preds of b
    for (auto& [b, blk] : blocks) {
        preds_rev[b] = blk.succs;
        succs_rev[b] = blk.preds;
    }
    std::vector<int64_t> exit_preds;
    for (auto& [b, blk] : blocks) {
        if (blk.succs.empty()) exit_preds.push_back(b);
    }

    std::set<int64_t> visited;
    std::vector<int64_t> order;
    std::function<void(int64_t)> dfs = [&](int64_t n) {
        visited.insert(n);
        const std::vector<int64_t>& nxt = (n == EXIT_MARKER) ? exit_preds : succs_rev[n];
        for (int64_t s : nxt) {
            if (!visited.count(s)) dfs(s);
        }
        order.push_back(n);
    };
    dfs(EXIT_MARKER);
    for (auto& [b, _] : blocks) {
        if (!visited.count(b)) {
            order.push_back(b);
            visited.insert(b);
        }
    }
    std::reverse(order.begin(), order.end());
    std::map<int64_t, size_t> rpo_index;
    for (size_t i = 0; i < order.size(); ++i) rpo_index[order[i]] = i;

    std::map<int64_t, int64_t> idom;
    idom[EXIT_MARKER] = EXIT_MARKER;
    bool changed = true;
    while (changed) {
        changed = false;
        for (int64_t b : order) {
            if (b == EXIT_MARKER) continue;
            std::vector<int64_t> ps = preds_rev.count(b) ? preds_rev[b] : std::vector<int64_t>{};
            if (std::find(exit_preds.begin(), exit_preds.end(), b) != exit_preds.end()) ps.push_back(EXIT_MARKER);
            std::vector<int64_t> ps_f;
            for (int64_t p : ps) if (idom.count(p)) ps_f.push_back(p);
            if (ps_f.empty()) continue;
            int64_t new_idom = ps_f[0];
            for (size_t k = 1; k < ps_f.size(); ++k) {
                int64_t a = new_idom, c = ps_f[k];
                while (a != c) {
                    while (rpo_index[a] > rpo_index[c]) a = idom[a];
                    while (rpo_index[c] > rpo_index[a]) c = idom[c];
                }
                new_idom = a;
            }
            auto it = idom.find(b);
            if (it == idom.end() || it->second != new_idom) {
                idom[b] = new_idom;
                changed = true;
            }
        }
    }
    std::map<int64_t, std::optional<int64_t>> result;
    for (auto& [b, _] : blocks) {
        auto it = idom.find(b);
        if (it == idom.end() || it->second == EXIT_MARKER) {
            result[b] = std::nullopt;
        } else {
            result[b] = it->second;
        }
    }
    return result;
}

}  // namespace nd
